File: pipeline/cobol_decimal.py
"""
COBOL numeric semantics — fixed-point decimal with PIC precision enforcement.

Faithfully reproduces COBOL arithmetic behavior:
- Silent left-truncation on overflow (no ON SIZE ERROR)
- Truncation of fractional digits (default) or ROUND_HALF_UP (ROUNDED phrase)
- COBOL-standard intermediate precision rules for ADD/SUB/MUL/DIV
- SPACES/blank coercion to zero for MOVE semantics
- Storage byte size computation for COMP, COMP-3, DISPLAY

IQ-09 UPGRADE NOTE: When IQ-09 (Differential Test Harness) is implemented,
skeleton integration MUST be upgraded to use CobolDecimal for all numeric fields.
See specs/iq-03-cobol-numeric-semantics/spec.md DD-06 for details.
"""

import math
from decimal import Decimal, ROUND_DOWN, ROUND_HALF_UP, InvalidOperation
from functools import total_ordering
from typing import Optional, Union


class CobolOverflowError(ArithmeticError):
    """Raised when on_size_error='raise' and value exceeds PIC capacity."""
    pass


@total_ordering
class CobolDecimal:
    """Fixed-point decimal with COBOL PIC semantics.

    Enforces exact precision, scale, sign, and overflow behavior matching
    the COBOL standard.

    Args:
        digits: Integer digit capacity (count of 9s before V in PIC).
        scale: Decimal places (count of 9s after V in PIC).
        signed: True if PIC has S prefix.
        usage: Storage format — 'DISPLAY', 'COMP', 'COMP-3', 'BINARY'.
        on_size_error: 'truncate' (default, COBOL standard) or 'raise'.
    """

    __slots__ = ('digits', 'scale', 'signed', 'usage', 'on_size_error', '_value')

    def __init__(
        self,
        digits: int = 9,
        scale: int = 0,
        signed: bool = True,
        usage: str = 'DISPLAY',
        on_size_error: str = 'truncate',
    ):
        self.digits = digits
        self.scale = scale
        self.signed = signed
        self.usage = usage.upper() if usage else 'DISPLAY'
        self.on_size_error = on_size_error
        self._value = Decimal('0')

    # ── Properties ──────────────────────────────────────────────────────

    @property
    def value(self) -> Decimal:
        return self._value

    @property
    def max_value(self) -> Decimal:
        """Maximum storable value based on PIC digits and scale."""
        int_part = Decimal('9' * self.digits) if self.digits > 0 else Decimal('0')
        if self.scale > 0:
            frac_part = Decimal('0.' + '9' * self.scale)
            return int_part + frac_part
        return int_part

    @property
    def min_value(self) -> Decimal:
        """Minimum storable value."""
        if self.signed:
            return -self.max_value
        return Decimal('0')

    @property
    def total_digits(self) -> int:
        """Total digit count (integer + fractional)."""
        return self.digits + self.scale

    @property
    def storage_bytes(self) -> int:
        """Byte size based on USAGE clause.

        Rules from COBOL standard:
        - DISPLAY: 1 byte per digit (sign embedded in last byte)
        - COMP-3 / PACKED-DECIMAL: ceil((total_digits + 1) / 2)
        - COMP / BINARY: 2 (≤4 digits), 4 (≤9), 8 (≤18)
        """
        td = self.total_digits
        usage = self.usage

        if usage in ('COMP-3', 'PACKED-DECIMAL'):
            return math.ceil((td + 1) / 2)
        elif usage in ('COMP', 'BINARY', 'COMP-4', 'COMP-5'):
            if td <= 4:
                return 2
            elif td <= 9:
                return 4
            else:
                return 8
        else:
            # DISPLAY: 1 byte per digit
            return td

    # ── Mutation ────────────────────────────────────────────────────────

    def set(
        self,
        value: Union[int, float, Decimal, str],
        rounded: bool = False,
    ) -> 'CobolDecimal':
        """Assign a value, enforcing PIC precision.

        Truncates or rounds fractional digits, then truncates left digits
        if the integer part exceeds the PIC capacity.

        Args:
            value: The value to store.
            rounded: If True, use ROUND_HALF_UP (COBOL ROUNDED phrase).
                     If False, truncate toward zero (COBOL default).
        Returns:
            self, for chaining.
        """
        d = Decimal(str(value))

        # Unsigned fields: take absolute value
        if not self.signed:
            d = abs(d)

        self._value = self._truncate_to_pic(d, rounded=rounded)
        return self

    def from_display(
        self,
        raw: Union[str, int, float, Decimal, None],
    ) -> 'CobolDecimal':
        """Coerce a display value into this field.

        Handles COBOL MOVE semantics:
        - SPACES / None / empty → 0
        - ZEROS → 0
        - Digit string → parse with implied decimal at scale position
        - Numeric types → direct conversion
        """
        if raw is None:
            self._value = Decimal('0')
            return self

        if isinstance(raw, (int, float)):
            return self.set(raw)

        if isinstance(raw, Decimal):
            return self.set(raw)

        # String handling
        s = str(raw).strip().upper()

        if s in ('', 'SPACES', 'SPACE', 'ZEROS', 'ZEROES', 'ZERO'):
            self._value = Decimal('0')
            return self

        # Digit string with implied decimal point
        # For PIC S9(5)V99, '1234567' → 12345.67 (last `scale` digits are fractional)
        s_clean = s.lstrip('+-')
        is_negative = s.startswith('-')

        if s_clean.isdigit() and self.scale > 0 and '.' not in s:
            # Implied decimal: last `scale` digits are fractional
            if len(s_clean) > self.scale:
                int_part = s_clean[:-self.scale]
                frac_part = s_clean[-self.scale:]
            else:
                int_part = '0'
                frac_part = s_clean.zfill(self.scale)
            d = Decimal(f"{int_part}.{frac_part}")
            if is_negative:
                d = -d
            return self.set(d)

        # Try parsing as a number directly
        try:
            return self.set(Decimal(s))
        except InvalidOperation:
            # Unparseable → zero (COBOL MOVE of non-numeric to numeric)
            self._value = Decimal('0')
            return self

    # ── Arithmetic (COBOL intermediate precision rules) ─────────────────

    def add(self, other: 'CobolDecimal') -> 'CobolDecimal':
        """ADD with COBOL intermediate precision.

        Intermediate: max(d1,d2)+1 integer digits, max(s1,s2) scale.
        """
        int_digits = max(self.digits, other.digits) + 1
        int_scale = max(self.scale, other.scale)
        result = CobolDecimal(digits=int_digits, scale=int_scale, signed=True)
        raw = self._value + other._value
        result._value = result._truncate_to_pic(raw)
        return result

    def subtract(self, other: 'CobolDecimal') -> 'CobolDecimal':
        """SUBTRACT with COBOL intermediate precision.

        Intermediate: max(d1,d2)+1 integer digits, max(s1,s2) scale.
        """
        int_digits = max(self.digits, other.digits) + 1
        int_scale = max(self.scale, other.scale)
        result = CobolDecimal(digits=int_digits, scale=int_scale, signed=True)
        raw = self._value - other._value
        result._value = result._truncate_to_pic(raw)
        return result

    def multiply(self, other: 'CobolDecimal') -> 'CobolDecimal':
        """MULTIPLY with COBOL intermediate precision.

        Intermediate integer digits: d1 + d2 (enough room for the product).
        Intermediate scale: max(s1, s2) — the COBOL standard specifies s1+s2
        as the theoretical maximum, but actual implementations truncate the
        intermediate to the receiving field's scale. Since the intermediate
        has no receiving field yet, we use the more precise operand's scale,
        which matches the result you'd get assigning into a field with that
        scale.
        """
        int_digits = self.digits + other.digits
        int_scale = max(self.scale, other.scale)
        result = CobolDecimal(digits=int_digits, scale=int_scale, signed=True)
        raw = self._value * other._value
        result._value = result._truncate_to_pic(raw, rounded=True)
        return result

    def divide(self, other: 'CobolDecimal') -> 'CobolDecimal':
        """DIVIDE with COBOL intermediate precision.

        Intermediate precision is implementation-defined in the COBOL standard.
        We use a conservative upper bound:
          integer digits = d1 + s2 + max(s1, s2)
          scale = max(s1, s2) + dividend scale
        This ensures the quotient has enough room for exact representation
        up to the receiving field's precision.
        """
        if other._value == 0:
            raise ZeroDivisionError("COBOL DIVIDE by zero")

        max_scale = max(self.scale, other.scale)
        int_digits = self.digits + other.scale + max_scale
        int_scale = max_scale + self.scale

        # Ensure minimum reasonable precision
        int_digits = max(int_digits, self.digits)
        int_scale = max(int_scale, self.scale)

        result = CobolDecimal(digits=int_digits, scale=int_scale, signed=True)
        raw = self._value / other._value
        result._value = result._truncate_to_pic(raw)
        return result

    def assign_to(
        self,
        target: 'CobolDecimal',
        rounded: bool = False,
    ) -> 'CobolDecimal':
        """Assign this value to a target CobolDecimal, truncating to target's PIC.

        This is the final step of a COBOL COMPUTE: the intermediate result
        is truncated (or rounded) to fit the receiving field.
        """
        target.set(self._value, rounded=rounded)
        return target

    # ── Internal ────────────────────────────────────────────────────────

    def _truncate_to_pic(self, d: Decimal, rounded: bool = False) -> Decimal:
        """Truncate a Decimal to fit this field's PIC definition.

        1. Truncate (or round) fractional digits to `self.scale`
        2. If integer part exceeds `self.digits`, truncate leftmost digits
        3. If on_size_error='raise', raise instead of truncating left
        """
        sign = Decimal('-1') if d < 0 else Decimal('1')
        abs_d = abs(d)

        # Step 1: Handle fractional digits
        if self.scale >= 0:
            quantize_exp = Decimal(10) ** -self.scale
            if rounded:
                abs_d = abs_d.quantize(quantize_exp, rounding=ROUND_HALF_UP)
            else:
                abs_d = abs_d.quantize(quantize_exp, rounding=ROUND_DOWN)

        # Step 2: Check integer part against digit capacity
        int_part = int(abs_d)
        max_int = 10 ** self.digits - 1 if self.digits > 0 else 0

        if int_part > max_int:
            if self.on_size_error == 'raise':
                raise CobolOverflowError(
                    f"Value {d} exceeds PIC capacity: "
                    f"{self.digits} integer digits (max {max_int})"
                )
            # Truncate left digits: keep only the rightmost `digits` digits
            int_part = int_part % (10 ** self.digits) if self.digits > 0 else 0
            # Reconstruct with fractional part preserved
            frac_part = abs_d - int(abs_d)
            abs_d = Decimal(int_part) + frac_part

        # Unsigned: already ensured positive in set()
        if not self.signed:
            return abs_d
        return abs_d * sign

    # ── Python protocol ─────────────────────────────────────────────────

    def __eq__(self, other: object) -> bool:
        if isinstance(other, CobolDecimal):
            return self._value == other._value
        if isinstance(other, (int, float, Decimal)):
            return self._value == Decimal(str(other))
        return NotImplemented

    def __lt__(self, other: object) -> bool:
        if isinstance(other, CobolDecimal):
            return self._value < other._value
        if isinstance(other, (int, float, Decimal)):
            return self._value < Decimal(str(other))
        return NotImplemented

    def __add__(self, other: 'CobolDecimal') -> 'CobolDecimal':
        return self.add(other)

    def __sub__(self, other: 'CobolDecimal') -> 'CobolDecimal':
        return self.subtract(other)

    def __mul__(self, other: 'CobolDecimal') -> 'CobolDecimal':
        return self.multiply(other)

    def __truediv__(self, other: 'CobolDecimal') -> 'CobolDecimal':
        return self.divide(other)

    def __repr__(self) -> str:
        sign = 'S' if self.signed else ''
        pic = f"{sign}9({self.digits})"
        if self.scale > 0:
            pic += f"V9({self.scale})"
        usage = f" {self.usage}" if self.usage != 'DISPLAY' else ''
        return f"CobolDecimal({pic}{usage}, value={self._value})"

    def __str__(self) -> str:
        return str(self._value)

    def __hash__(self) -> int:
        return hash(self._value)

File: pipeline/reimpl/java/cobol-decimal/src/main/java/com/modernization/masquerade/cobol/CobolDecimal.java
package com.modernization.masquerade.cobol;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

/**
 * Fixed-point decimal with COBOL PIC semantics — Java port of
 * {@code pipeline/cobol_decimal.py}.
 *
 * <p>Faithfully reproduces COBOL arithmetic behavior:
 * <ul>
 *   <li>Silent left-truncation on overflow (no ON SIZE ERROR)</li>
 *   <li>Truncation of fractional digits (default) or ROUND_HALF_UP (ROUNDED phrase)</li>
 *   <li>COBOL-standard intermediate precision rules for ADD/SUB/MUL/DIV</li>
 *   <li>SPACES/blank coercion to zero for MOVE semantics</li>
 *   <li>Storage byte size computation for COMP, COMP-3, DISPLAY</li>
 * </ul>
 *
 * <p><b>Parity contract:</b> every test vector in {@code CobolDecimalTest} must
 * produce byte-identical results to the equivalent vector in
 * {@code pipeline/tests/test_cobol_decimal.py}. Drift between this class and
 * the Python original is a correctness bug.
 */
public final class CobolDecimal implements Comparable<CobolDecimal> {

    /** Behavior when an assigned value exceeds the integer-digit capacity. */
    public enum OnSizeError {
        /** Silently drop the leftmost digits that exceed PIC capacity (COBOL default). */
        TRUNCATE,
        /** Throw {@link CobolOverflowError}. */
        RAISE
    }

    private final int digits;
    private final int scale;
    private final boolean signed;
    private final String usage;
    private final OnSizeError onSizeError;
    private BigDecimal value;

    // ── Constructors ────────────────────────────────────────────────────

    /**
     * @param digits      Integer digit capacity (count of 9s before V in PIC).
     * @param scale       Decimal places (count of 9s after V in PIC).
     * @param signed      True if PIC has S prefix.
     * @param usage       Storage format — "DISPLAY", "COMP", "COMP-3", "BINARY".
     * @param onSizeError TRUNCATE (default, COBOL standard) or RAISE.
     */
    public CobolDecimal(int digits, int scale, boolean signed, String usage, OnSizeError onSizeError) {
        this.digits = digits;
        this.scale = scale;
        this.signed = signed;
        this.usage = usage == null ? "DISPLAY" : usage.toUpperCase();
        this.onSizeError = onSizeError == null ? OnSizeError.TRUNCATE : onSizeError;
        this.value = BigDecimal.ZERO;
    }

    public CobolDecimal(int digits, int scale, boolean signed) {
        this(digits, scale, signed, "DISPLAY", OnSizeError.TRUNCATE);
    }

    public CobolDecimal(int digits, int scale, boolean signed, String usage) {
        this(digits, scale, signed, usage, OnSizeError.TRUNCATE);
    }

    // ── Properties ──────────────────────────────────────────────────────

    public BigDecimal value() { return value; }
    public int digits() { return digits; }
    public int scale() { return scale; }
    public boolean signed() { return signed; }
    public String usage() { return usage; }
    public OnSizeError onSizeError() { return onSizeError; }

    /** Maximum storable value based on PIC digits and scale. */
    public BigDecimal maxValue() {
        BigDecimal intPart = digits > 0 ? new BigDecimal("9".repeat(digits)) : BigDecimal.ZERO;
        if (scale > 0) {
            BigDecimal fracPart = new BigDecimal("0." + "9".repeat(scale));
            return intPart.add(fracPart);
        }
        return intPart;
    }

    /** Minimum storable value (negative of max if signed, else zero). */
    public BigDecimal minValue() {
        return signed ? maxValue().negate() : BigDecimal.ZERO;
    }

    /** Total digit count (integer + fractional). */
    public int totalDigits() {
        return digits + scale;
    }

    /**
     * Byte size based on USAGE clause. Mirrors {@code cobol_decimal.py:storage_bytes}.
     * <ul>
     *   <li>DISPLAY: 1 byte per digit (sign embedded in last byte)</li>
     *   <li>COMP-3 / PACKED-DECIMAL: ceil((total_digits + 1) / 2)</li>
     *   <li>COMP / BINARY: 2 (≤4 digits), 4 (≤9), 8 (≤18)</li>
     * </ul>
     */
    public int storageBytes() {
        int td = totalDigits();
        switch (usage) {
            case "COMP-3":
            case "PACKED-DECIMAL":
                // ceil((td + 1) / 2) — written as integer math
                return (td + 1 + 1) / 2;
            case "COMP":
            case "BINARY":
            case "COMP-4":
            case "COMP-5":
                if (td <= 4) return 2;
                if (td <= 9) return 4;
                return 8;
            default:
                return td;
        }
    }

    // ── Mutation ────────────────────────────────────────────────────────

    /** Assign a value, enforcing PIC precision. Truncates fractional digits. */
    public CobolDecimal set(BigDecimal v) {
        return set(v, false);
    }

    /**
     * Assign a value, enforcing PIC precision.
     *
     * @param v       The value to store.
     * @param rounded If true, use ROUND_HALF_UP (COBOL ROUNDED phrase).
     *                If false, truncate toward zero (COBOL default).
     */
    public CobolDecimal set(BigDecimal v, boolean rounded) {
        BigDecimal d = signed ? v : v.abs();
        this.value = truncateToPic(d, rounded);
        return this;
    }

    public CobolDecimal set(long v) {
        return set(BigDecimal.valueOf(v));
    }

    public CobolDecimal set(long v, boolean rounded) {
        return set(BigDecimal.valueOf(v), rounded);
    }

    public CobolDecimal set(String v) {
        return set(new BigDecimal(v));
    }

    /**
     * Coerce a display value into this field. Mirrors {@code cobol_decimal.py:from_display}.
     *
     * <p>Handles COBOL MOVE semantics:
     * <ul>
     *   <li>SPACES / null / empty → 0</li>
     *   <li>ZEROS → 0</li>
     *   <li>Digit string → parsed with implied decimal at scale position</li>
     *   <li>Numeric types → direct conversion</li>
     * </ul>
     */
    public CobolDecimal fromDisplay(Object raw) {
        if (raw == null) {
            this.value = BigDecimal.ZERO;
            return this;
        }
        if (raw instanceof BigDecimal bd) {
            return set(bd);
        }
        if (raw instanceof Number n) {
            return set(new BigDecimal(n.toString()));
        }

        String s = raw.toString().trim().toUpperCase();

        if (s.isEmpty()
                || s.equals("SPACES") || s.equals("SPACE")
                || s.equals("ZEROS") || s.equals("ZEROES") || s.equals("ZERO")) {
            this.value = BigDecimal.ZERO;
            return this;
        }

        // Digit string with implied decimal point.
        // For PIC S9(5)V99, '1234567' → 12345.67 (last `scale` digits are fractional).
        boolean isNegative = s.startsWith("-");
        String sClean = s.replaceAll("^[+-]", "");

        if (sClean.matches("\\d+") && scale > 0 && !s.contains(".")) {
            String intPart;
            String fracPart;
            if (sClean.length() > scale) {
                intPart = sClean.substring(0, sClean.length() - scale);
                fracPart = sClean.substring(sClean.length() - scale);
            } else {
                intPart = "0";
                // Left-pad with zeros to `scale` width
                StringBuilder padded = new StringBuilder();
                for (int i = 0; i < scale - sClean.length(); i++) padded.append('0');
                padded.append(sClean);
                fracPart = padded.toString();
            }
            BigDecimal d = new BigDecimal(intPart + "." + fracPart);
            if (isNegative) d = d.negate();
            return set(d);
        }

        try {
            return set(new BigDecimal(s));
        } catch (NumberFormatException e) {
            // Unparseable → zero (COBOL MOVE of non-numeric to numeric)
            this.value = BigDecimal.ZERO;
            return this;
        }
    }

    // ── Arithmetic (COBOL intermediate precision rules) ─────────────────

    /**
     * ADD with COBOL intermediate precision.
     * Intermediate: max(d1,d2)+1 integer digits, max(s1,s2) scale.
     */
    public CobolDecimal add(CobolDecimal other) {
        int intDigits = Math.max(this.digits, other.digits) + 1;
        int intScale = Math.max(this.scale, other.scale);
        CobolDecimal result = new CobolDecimal(intDigits, intScale, true);
        BigDecimal raw = this.value.add(other.value);
        result.value = result.truncateToPic(raw, false);
        return result;
    }

    /**
     * SUBTRACT with COBOL intermediate precision (same rule as ADD).
     */
    public CobolDecimal subtract(CobolDecimal other) {
        int intDigits = Math.max(this.digits, other.digits) + 1;
        int intScale = Math.max(this.scale, other.scale);
        CobolDecimal result = new CobolDecimal(intDigits, intScale, true);
        BigDecimal raw = this.value.subtract(other.value);
        result.value = result.truncateToPic(raw, false);
        return result;
    }

    /**
     * MULTIPLY with COBOL intermediate precision.
     * Intermediate integer digits: d1 + d2. Intermediate scale: max(s1, s2).
     * Multiplication uses {@code rounded=true} in the truncate step to match
     * the Python original.
     */
    public CobolDecimal multiply(CobolDecimal other) {
        int intDigits = this.digits + other.digits;
        int intScale = Math.max(this.scale, other.scale);
        CobolDecimal result = new CobolDecimal(intDigits, intScale, true);
        BigDecimal raw = this.value.multiply(other.value);
        result.value = result.truncateToPic(raw, true);
        return result;
    }

    /**
     * DIVIDE with conservative intermediate precision.
     * Intermediate integer digits = d1 + s2 + max(s1, s2).
     * Intermediate scale = max(s1, s2) + dividend scale.
     */
    public CobolDecimal divide(CobolDecimal other) {
        if (other.value.signum() == 0) {
            throw new ArithmeticException("COBOL DIVIDE by zero");
        }

        int maxScale = Math.max(this.scale, other.scale);
        int intDigits = this.digits + other.scale + maxScale;
        int intScale = maxScale + this.scale;

        // Ensure minimum reasonable precision
        intDigits = Math.max(intDigits, this.digits);
        intScale = Math.max(intScale, this.scale);

        CobolDecimal result = new CobolDecimal(intDigits, intScale, true);
        // BigDecimal.divide requires explicit scale + rounding for non-terminating quotients.
        // Use a generous buffer beyond intScale so the subsequent truncateToPic has room.
        BigDecimal raw = this.value.divide(other.value, intScale + 10, RoundingMode.DOWN);
        result.value = result.truncateToPic(raw, false);
        return result;
    }

    /**
     * Assign this value to a target CobolDecimal, truncating to the target's PIC.
     * This is the final step of a COBOL COMPUTE: the intermediate result is
     * truncated (or rounded) to fit the receiving field.
     */
    public CobolDecimal assignTo(CobolDecimal target) {
        return assignTo(target, false);
    }

    public CobolDecimal assignTo(CobolDecimal target, boolean rounded) {
        target.set(this.value, rounded);
        return target;
    }

    // ── Internal ────────────────────────────────────────────────────────

    /**
     * Truncate a BigDecimal to fit this field's PIC definition.
     * <ol>
     *   <li>Truncate (or round) fractional digits to {@code this.scale}</li>
     *   <li>If integer part exceeds {@code this.digits}, truncate leftmost digits</li>
     *   <li>If onSizeError == RAISE, raise instead of truncating left</li>
     * </ol>
     */
    private BigDecimal truncateToPic(BigDecimal d, boolean rounded) {
        int sign = d.signum() < 0 ? -1 : 1;
        BigDecimal absD = d.abs();

        // Step 1: Handle fractional digits
        if (scale >= 0) {
            absD = absD.setScale(scale, rounded ? RoundingMode.HALF_UP : RoundingMode.DOWN);
        }

        // Step 2: Check integer part against digit capacity
        BigDecimal intPart = absD.setScale(0, RoundingMode.DOWN);
        BigDecimal maxInt = digits > 0
                ? BigDecimal.TEN.pow(digits).subtract(BigDecimal.ONE)
                : BigDecimal.ZERO;

        if (intPart.compareTo(maxInt) > 0) {
            if (onSizeError == OnSizeError.RAISE) {
                throw new CobolOverflowError(
                        "Value " + d + " exceeds PIC capacity: "
                                + digits + " integer digits (max " + maxInt + ")");
            }
            // Truncate left digits: keep only the rightmost `digits` digits.
            // intPart % (10^digits), then re-attach the original fractional part.
            BigDecimal modulus = digits > 0 ? BigDecimal.TEN.pow(digits) : BigDecimal.ONE;
            BigDecimal intMod = digits > 0 ? intPart.remainder(modulus) : BigDecimal.ZERO;
            BigDecimal fracPart = absD.subtract(intPart);
            absD = intMod.add(fracPart);
        }

        if (!signed) return absD;
        return sign < 0 ? absD.negate() : absD;
    }

    // ── Comparable / equals / hashCode ──────────────────────────────────

    @Override
    public int compareTo(CobolDecimal other) {
        return this.value.compareTo(other.value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o instanceof CobolDecimal cd) {
            return this.value.compareTo(cd.value) == 0;
        }
        if (o instanceof BigDecimal bd) {
            return this.value.compareTo(bd) == 0;
        }
        if (o instanceof Number n) {
            return this.value.compareTo(new BigDecimal(n.toString())) == 0;
        }
        return false;
    }

    @Override
    public int hashCode() {
        // stripTrailingZeros so equal numeric values share a hash, matching the
        // Python behavior of hash(self._value) where Decimal('2.00') and
        // Decimal('2') hash equally.
        return Objects.hashCode(this.value.stripTrailingZeros());
    }

    @Override
    public String toString() {
        return value.toPlainString();
    }

    /** Debug rendering similar to {@code __repr__} on the Python class. */
    public String toDebugString() {
        String signPrefix = signed ? "S" : "";
        String pic = signPrefix + "9(" + digits + ")";
        if (scale > 0) pic += "V9(" + scale + ")";
        String u = !"DISPLAY".equals(usage) ? " " + usage : "";
        return "CobolDecimal(" + pic + u + ", value=" + value + ")";
    }
}
